### ISSUE TITLE
`oc start-build` conflicts with `contextDir` resulting in build error

### ISSUE THREAD

**Issue by quintesse** [openshift/origin#21646]
TL;DR Using `oc start-build --from-file path/to/binary` together with a BuildConfig that uses `contextDir` fails during the build process with `error: provided context directory does not exist: my_context_dir`. Somehow the uploaded binary gets validated against the context dir folder value, but that doesn't seem to make any sense.

So we're using a project with a BuildConfig that gets it's original code from a GitHub project.
But then during development we'll override that by executing "oc start-build" commands in our local development environment.

This works pretty well. In Java projects we can even choose to either push the source changes with the "--from-dir" option and let the BuildConfig/S2I do the building or we compile locally and use "--from-file" to just send the resulting JAR file.

So far so good. But I think I've encountered a problem when using this together with the "contextDir" value (in BuildConfig.spec.source). Because we have a Git repo that has multiple folders with sources, each for a different BuildConfig. The "contextDir" works fine for that, but the problem appears when trying to use it together with the "oc start-build" command.

Using "oc start-build testbuild --from-file path/to/jarfile.jar" in combination with a "contextDir" I get the following error in the logs for the "testbuild" build on OpenShift:

> Receiving source from STDIN as file jarfile.jar
> error: provided context directory does not exist: my_context_dir

Somehow it tries to validate the file I'm sending against the "contextDir" value which just doesn't seem to make any sense.

NB: The "--from-dir" option gives me the same problem if I try to run it with a path directly to the context dir, but at least there I can work around it and pass it the context dir's parent path. But I have the feeling that's incorrect too.

##### Version

oc v3.10.0+dd10d17
kubernetes v1.10.0+b81c8f8
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://192.168.42.87:8443
kubernetes v1.11.0+d4cacc0

##### Steps To Reproduce

- Have a BuildConfig similar to this:

```
  - kind: BuildConfig
    apiVersion: v1
    spec:
      source:
        type: Git
        git:
          uri: https://github.com/quintesse/generated-app-ex
        contextDir: backend
      strategy:
        type: Source
        sourceStrategy:
          from:
            kind: ImageStreamTag
            name: 'openjdk18-openshift:latest'
```

- Code obtained from a GitHub repo, for example: https://github.com/quintesse/generated-app-ex
- Clone locally
- In "backend" folder run `mvn package`
- Run `oc start-build -from-file target/my-app-1.0.0.jar `
- Observe error in OpenShift logs

If one were to move the code in te repository from the "backend" folder to the root and remove the `contextDir` property from the yaml then these steps work fine.

##### Current Result

> Receiving source from STDIN as file jarfile.jar
> error: provided context directory does not exist: my_context_dir

##### Expected Result

A working build, no error

**Comment by adambkaplan** [comment]
@openshift/sig-developer-experience

**Comment by bparees** [comment]
this is working as designed.

Regardless of where your input content comes from (a git repo or a binary input), the expectation is that "contextDir" is a valid path within the input.

this flow is described here:
https://docs.okd.io/latest/dev_guide/builds/build_inputs.html#how-build-inputs-work

If you want to supply content that does not contain your "contextDir" you'll have to create a second buildconfig for that purpose.

**Comment by quintesse** [comment]
For others encountering this same issue, this is the work-around we found: we create a temporary directory, create a sub folder structure the same as defined by `contextDir` and then copy the file we wanted to send into it. We then use `--from-dir` on that temporary directory.

It would be nice if `oc --from-file` could do that for us but there's a work-around at least, that's sufficient for our case.
